### PR TITLE
Add IgmpV3 support

### PR DESCRIPTION
**Pull request by jlshuman1961**
Added support for IGMP version 3 messages.

### CODE REVIEW

**Comment by jlshuman1961** [comment]
I found a shortcoming with the code I submitted in the Pull Request on made on Wednesday. How can I submit some changes to the Pull Request? Or do I need to make a new branch and submit a whole new Pull Request and just close the previous Pull Request?

**Comment by PhyxionNL** [comment]
> I found a shortcoming with the code I submitted in the Pull Request on made on Wednesday. How can I submit some changes to the Pull Request? Or do I need to make a new branch and submit a whole new Pull Request and just close the previous Pull Request?

You can simply commit your changes to your own repository and branch (https://github.com/jlshuman1961/packetnet/tree/IGMPv3-support) and it will be updated here automatically.

**Comment by jlshuman1961** [comment]
> 

Thanks, PhyxionNL, I made 5 more commits to fix the shortcoming that I mentioned. Please let me know if did correctly, or not.

File: PacketDotNet/IgmpV3MembershipQueryPacket.cs [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+                    this.sourceAddresses = new List<IPAddress>();
+                    var offset = Header.Offset + IgmpV3MembershipQueryFields.SourceAddressStart;
+
+                    for (int i = 0; i < NumberOfSources; i++)
+                    {
+                        long address = EndianBitConverter.Little.ToUInt32(Header.Bytes, offset);
+                        this.sourceAddresses.Add(new IPAddress(address));
+                        offset += 4;
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (jlshuman1961) [reply PacketDotNet/IgmpV3MembershipQueryPacket.cs:158]: I can add a setter, but I don't understand why you would need one. The purpose of this code is to retrieve the list of source addresses from an IGMP message in a packet. Why would someone want to change the actual data from a packet? Is this code used to create new IGMP messages? If so, then I can understand having a setter.

[the remark this answers is not in the document]
Comment (PhyxionNL) [reply PacketDotNet/IgmpV3MembershipQueryPacket.cs:158]: All packets in PDN can be created/modified. PDN is not just used to monitor/read packets, it's also used to modify packets or inject new packets 👍

[the remark this answers is not in the document]
Comment (jlshuman1961) [reply PacketDotNet/IgmpV3MembershipQueryPacket.cs:158]: When I make these changes, do I commit them the same way I made my second update?

[the remark this answers is not in the document]
Comment (PhyxionNL) [reply PacketDotNet/IgmpV3MembershipQueryPacket.cs:158]: Yes please :)

[the remark this answers is not in the document]
Comment (jlshuman1961) [reply PacketDotNet/IgmpV3MembershipQueryPacket.cs:158]: In IgmpV2Packet.cs, the field called, "GroupAddress" is not a property with a setter. It is read only. Why is that? I replicated this in IgmpV3MemvershipQueryPacket. How is a user supposed to set the GroupAddress in these cases?

[the remark this answers is not in the document]
Comment (PhyxionNL) [reply PacketDotNet/IgmpV3MembershipQueryPacket.cs:158]: Very close to the get, see https://github.com/dotpcap/packetnet/blob/master/PacketDotNet/IPv4Packet.cs#L290 for an example. You basically write to the same offset as you read it from.

[the remark this answers is not in the document]
Comment (jlshuman1961) [reply PacketDotNet/IgmpV3MembershipQueryPacket.cs]: Committed new code for this section.

File: PacketDotNet/IgmpV3MembershipReportGroupRecord.cs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+{
+    /// <summary>
+    /// IGMP v3 Membership Report Group Record.
+    /// </summary>
+    public class IgmpV3MembershipReportGroupRecord
+    {
+        /// <summary>
+        /// ByteArraySegment representing the header passed in to the Construtor.
+        /// </summary>
+        private ByteArraySegment header;
```
[the remark this answers is not in the document]
Comment (jlshuman1961) [reply PacketDotNet/IgmpV3MembershipReportGroupRecord.cs]: Committed new code for this section.

File: PacketDotNet/IgmpV3MembershipReportGroupRecord.cs [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+                    for (int i = 0; i < NumberOfSources; i++)
+                    {
+                        long address = EndianBitConverter.Little.ToUInt32(this.header.Bytes, offset);
+                        this.sourceAddresses.Add(new IPAddress(address));
+                        offset += 4;
+                    }
+                }
+
+                return this.sourceAddresses;
```
[the remark this answers is not in the document]
Comment (jlshuman1961) [reply PacketDotNet/IgmpV3MembershipReportGroupRecord.cs]: Committed new code for this section.

File: PacketDotNet/IgmpV3MembershipReportPacket.cs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+                    for (int i = 0; i < NumberOfGroupRecords; i++)
+                    {
+                        IgmpV3MembershipReportGroupRecord groupRecord = new IgmpV3MembershipReportGroupRecord(Header, offset);
+                        this.groupRecords.Add(groupRecord);
+                        offset += (IgmpV3MembershipReportGroupRecordFields.IgmpV3MembershipReportGroupRecordHeaderLength + groupRecord.NumberOfSources * IPv4Fields.AddressLength);
+                    }
+                }
+
+                return this.groupRecords;
```
[the remark this answers is not in the document]
Comment (jlshuman1961) [reply PacketDotNet/IgmpV3MembershipReportPacket.cs]: Committed new code for this section.

File: README.md [hunk 5]
```diff
@@ -3,8 +3,7 @@
 
 New!
 ==========
-The newly released [PacketDotNet.Connections](http://github.com/chmorgan/packetnet-connections) is a supplement to Packet.NET that
-adds connection tracking and http following support.
+Added support for IGMP version 3.
 
 Packet.Net
 ==========
```
[the remark this answers is not in the document]
Comment (jlshuman1961) [reply README.md]: Fixed as requested.

**Comment by jlshuman1961** [comment]
Made changes as requested.

**Comment by jlshuman1961** [comment]
Made changes as requested.

**Comment by PhyxionNL** [comment]
> Made changes as requested.

Build fails.

**Comment by jlshuman1961** [comment]
I forgot to add using statement for my previous code submission. Added those in.

**Comment by jlshuman1961** [comment]
I don't know what to do about this build error. Something about not installing winpcap. Did I break something to make that happen?

**Comment by PhyxionNL** [comment]
> I don't know what to do about this build error. Something about not installing winpcap. Did I break something to make that happen?

No, Windows fails at the moment due to https://github.com/dotpcap/packetnet/issues/150. Doesn't matter for this PR. It looks good now, I'll format the documents and merge it.

**Comment by jlshuman1961** [comment]
How long will it be before we can see this IGMPv3 support in an official release of PacketDotNet?

**Comment by jlshuman1961** [comment]
I just noticed that there are still two properties that may need setters in them. This is in the IgmpV3MembershipQueryPacket.cs file. The properties are MaxResponseTime (a calculated value based on MaxResponseCode) and QueriersQueryInterval (a calculated value based on QueriersQueryIntervalCode). Or, I need to make MaxResponseCode and QueriersQueryIntervalCode public, so that the user can set those bytes directly. This is an oversight on my part because it is code that I would not normally use.
What do you want me to do about this?

**Comment by PhyxionNL** [comment]
If you can add setters to these properties, that'd be great. You can submit a follow up PR. Please start fresh from master branch here.

**Pull request merged**